### ISSUE TITLE
helloworld and ping cannot run same time in container

### ISSUE THREAD

**Issue by 9606** [F-Stack/f-stack#455]
Hi, f-stack team, 
I am writing a ping program based helloworld. I want to first run helloworld as DPDK PRIMARY process and then start ping program as SECONDARY process , then send ICMP packets through the ping program. 
If the interface is physical NIC，there is no problem， but if the interface is vhost-user，the ping program cannot work。
The test commands is both:
./helloworld -c /data/f-stack/config.ini
./ping -c /data/f-stack/config.ini x.x.x.x
the ping code is in https://github.com/9606/f-stack/commit/b75bbe49d5d3decf95d9792bd68c989ba91f2292 

the physical NIC config:
```
port_list=0

# Number of vdev.
nb_vdev=0

# Number of bond.
nb_bond=0

# Port config section
# Correspond to dpdk.port_list's index: port0, port1...
[port0]
addr=192.168.10.10
netmask=255.255.225.0
broadcast=192.168.10.255
gateway=192.168.10.1

# lcore list used to handle this port
# the format is same as port_list
lcore_list=0
```
the vhost-user config:
```
port_list=0

# Number of vdev.
nb_vdev=1

# Number of bond.
nb_bond=0

# Port config section
# Correspond to dpdk.port_list's index: port0, port1...
[port0]
addr=192.168.1.2
netmask=255.255.225.252
broadcast=192.168.1.3
gateway=192.168.1.1

# lcore list used to handle this port
# the format is same as port_list
#lcore_list=0

# bonding slave port list used to handle this port
# need to config while this port is a bonding port
# the format is same as port_list
#slave_port_list=0,1

# Packet capture path, this will hurt performance
#pcap=./a.pcap

[vdev0]
##iface=/usr/local/var/run/openvswitch/vhost-user0
path=/tmp/eth_vhost1.sock
queues=1
queue_size=256
mac=2a:df:a1:41:cc:85
cq=1
```

How can I solve this problem?

Thanks!

**Comment by jfb8856606** [comment]
If the helloworld can work correctly on docker?

**Comment by 9606** [comment]
Yes, helloworld and ping now can work well together in container. I fix the problem in https://github.com/9606/f-stack/commit/7168e6dabfb995d125ff2dd1b7e9d53c6a0ab25f. But I don't know if this is the proper way to write a ping program. And I don't understand why physical NICs don't need to add pipeline_dispatch_cb method. @jfb8856606

**Comment by 9606** [comment]
> If the helloworld can work correctly on docker?

Yes, helloworld and ping now can work well together in container. I fix the problem in https://github.com/9606/f-stack/commit/7168e6dabfb995d125ff2dd1b7e9d53c6a0ab25f. But I don't know if this is the proper way to write a ping program. And I don't understand why physical NICs don't need to add pipeline_dispatch_cb method. @jfb8856606

**Comment by jfb8856606** [comment]
I think the reason is `ff_rss_check` can't work well use vhost-user, but work well use physical NICs. Another reason may be `ff_rss_check` can't work both vhost-user and physical NICs, just work well accidental.

I will test it later.

**Comment by 9606** [comment]
Hello, I recently tested again. There is no problem with ovs + dpdk. When I change another data plane + dpdk, there has problems, so the reason why it can't work before should be the data plane I use. 
However, I think it is not good to write ping program in this way. Ping and curl should share the protocol stack that was started before, so I added some socket api through ipc mode and successfully ported the freebsd ping program to fstack, but there has copy of the parameters now. I will write some more code recently. Are you interested in merging? @jfb8856606

**Comment by jfb8856606** [comment]
@9606 Yes，looking forward to seeing your pull-request.

**Comment by piaoxuebingfeng** [comment]
@9606 when i compile ping.c ，i get some errors:
/f-stack/dpdk/x86_64-native-linuxapp-gcc/include/rte_memcpy.h:793:2: error: incompatible type for argument 2 of ‘_mm_storeu_si128’
  MOVEUNALIGNED_LEFT47(dst, src, n, srcofs);
  ^
In file included from /usr/lib/gcc/x86_64-redhat-linux/4.8.5/include/xmmintrin.h:1246:0,
                 from /usr/lib/gcc/x86_64-redhat-linux/4.8.5/include/x86intrin.h:34,
                 from /home/qiao/9606/f-stack/dpdk/x86_64-native-linuxapp-gcc/include/rte_vect.h:28,
                 from /home/qiao/9606/f-stack/dpdk/x86_64-native-linuxapp-gcc/include/rte_memcpy.h:17,
                 from /home/qiao/9606/f-stack/dpdk/x86_64-native-linuxapp-gcc/include/rte_ether.h:21,
                 from nse.c:12:
/usr/lib/gcc/x86_64-redhat-linux/4.8.5/include/emmintrin.h:700:1: note: expected ‘__m128i’ but argument is of type ‘int’
 _mm_storeu_si128 (__m128i *__P, __m128i __B)
 ^
make: *** [all] Error 1


how can i resolve it?

**Comment by jfb8856606** [comment]
Thanks for the detailed report and the follow-up!

To summarize the resolution:
- The original issue (ping secondary process not receiving packets over vhost-user) was resolved by adding a `pipeline_dispatch_cb` in the secondary process. As @jfb8856606 noted, the underlying reason is that `ff_rss_check` does not work reliably with vhost-user (which has no hardware RSS), so packets may not be correctly dispatched to the secondary process's lcore without explicit dispatch logic.
- With OVS+DPDK, the issue was not reproducible.

The IPC-based FreeBSD ping port mentioned in the last comment sounds interesting, but since no PR has been submitted in 5+ years, we'll close this issue for now. If anyone has a patch or wants to revive the discussion, please open a new issue or PR.

Closing as resolved.
